### ISSUE TITLE
feature request: RPC interface or intermediate representation API

### ISSUE THREAD

**Issue by gonzojive** [google/jax#2846]
I would like to use JAX from Go. Mainly I'm interested in the automatic differentiation feature in particular, otherwise it seems I could use XLA directly. Is there an RPC layer on top of JAX or intermediate representation that my Go program could generate?

**Comment by froystig** [comment]
We have neither interface currently. Internally, we have an intermediate representation that we call the [Jaxpr](https://jax.readthedocs.io/en/latest/jaxpr.html) IR, but we don't consider it stable at the moment.

If you have an IR in mind, one approach is to apply JAX to a Python function that interprets it using `jax.numpy` (or other JAX-compatible libraries). Here's a sketch of such an interpreter:

```python
import jax.numpy as jnp

def ir_call(ir_func, py_args):
  env = {arg: val for arg, val in zip(ir_func.args, py_args)}
  for op, dest, operands in ir_func.statements:
    if op == 'ADD':
      x, y = operands
      env[dest] = jnp.add(env[x], env[y])
    # ...
  return env[ir_func.out]
```

Once bound to a particular `ir_func`, the above is a Python function compatible with JAX's transformations, including autodiff. Continuing the example, to compute the gradient of an IR-defined function at the origin:

```python
def ir_grad_at(ir_func, *args):
  py_func = functools.partial(ir_call, ir_func)
  return jax.grad(py_func)(*args)

ir_func = ir_parse(some_text)
origin = [np.zeros(shape) for shape in ir_func.arg_shapes]
print(ir_grad_at(ir_func, *origin))
```

A more complete example is [this onnx interpreter](https://github.com/google/jax/blob/master/examples/onnx2xla.py) in our examples directory.

This approach works to evaluate a function's derivatives _at a given point_, or several. Conceptually, would something like this work for you?

**Comment by mattjj** [comment]
I think this is out of scope.
